Improve perf for forward AD view handling

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57058 Remove unused forward AD flag
* **#57057 Improve perf for forward AD view handling**
* #57152 fix InplaceOrView implementation for manual functions
* #57069 Fix use of allow_tensor_metadata in view variable creation

This PR performs optimization on the ViewInfo handling to remove the need for the "no forward AD mode".
- When the forward and backward ViewInfo are the same, create and store only one of them


Code for timing:
```python
timer = Timer(
    stmt='a.view(-1)',
    setup='''\
import torch
a = torch.rand(4)''')

res = timer.collect_callgrind(repeats=2, number=10)[1]
```

Difference between master and this PR:
```
# Benchmark at master
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33be83690>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69286                      68442
    Baseline:             1332                       1188
10 runs per measurement, 1 thread

# Benchmark at this branch
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33bd7ec30>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69437                      68562
    Baseline:             1363                       1188
10 runs per measurement, 1 thread

# Difference between the two
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fe1216e9a00>
    160  ???:0x000000000a11c8d0
     60  torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta
     60  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
     40  ???:0x0000000008e14f50
     40  ???:0x0000000008e05bd0
     40  ???:0x0000000008e05480
     40  ???:0x0000000008e036d0
     40  ???:0x0000000008e02720
     30  make_variable_differentiable_view
    ...
    -20  ???:0x0000000008e02060
    -20  ???:0x0000000008e01fd0
    -30  ???:torch::autograd::isForwardADEnabled()
    -40  ???:0x0000000008e14f90
    -40  ???:0x0000000008e05c00
    -40  ???:0x0000000008e054a0
    -40  ???:0x0000000008e036f0
    -40  ???:0x0000000008e02740
   -160  ???:0x000000000a11d8d0

Total: 120

```

Differential Revision: [D28071505](https://our.internmc.facebook.com/intern/diff/D28071505)

albanD committed:
Improve perf for forward AD view handling

[ghstack-poisoned]

albanD committed:
Update on "Improve perf for forward AD view handling"


This PR performs optimization on the ViewInfo handling to remove the need for the "no forward AD mode".
- When the forward and backward ViewInfo are the same, create and store only one of them


Code for timing:
```python
timer = Timer(
    stmt='a.view(-1)',
    setup='''\
import torch
a = torch.rand(4)''')

res = timer.collect_callgrind(repeats=2, number=10)[1]
```

Difference between master and this PR:
```
# Benchmark at master
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33be83690>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69286                      68442
    Baseline:             1332                       1188
10 runs per measurement, 1 thread

# Benchmark at this branch
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33bd7ec30>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69437                      68562
    Baseline:             1363                       1188
10 runs per measurement, 1 thread

# Difference between the two
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fe1216e9a00>
    160  ???:0x000000000a11c8d0
     60  torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta
     60  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
     40  ???:0x0000000008e14f50
     40  ???:0x0000000008e05bd0
     40  ???:0x0000000008e05480
     40  ???:0x0000000008e036d0
     40  ???:0x0000000008e02720
     30  make_variable_differentiable_view
    ...
    -20  ???:0x0000000008e02060
    -20  ???:0x0000000008e01fd0
    -30  ???:torch::autograd::isForwardADEnabled()
    -40  ???:0x0000000008e14f90
    -40  ???:0x0000000008e05c00
    -40  ???:0x0000000008e054a0
    -40  ???:0x0000000008e036f0
    -40  ???:0x0000000008e02740
   -160  ???:0x000000000a11d8d0

Total: 120

```


[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -146,33 +146,26 @@ inline Tensor as_view(const Tensor & base, const Tensor & tensor, bool is_bw_dif
+
+  auto diff_view_meta = torch::autograd::impl::get_view_autograd_meta(base);
+
+  // Special case when view info can be shared for forward and backward differentiable views
```
Is it a special case? Isn't this supposed to be the common case?

albanD replied:
It is the common case indeed.
I meant special case compared to the code that comes after that could handle all the cases.

Let me update the comment to make that distinction clearer.

albanD committed:
Update on "Improve perf for forward AD view handling"


This PR performs optimization on the ViewInfo handling to remove the need for the "no forward AD mode".
- When the forward and backward ViewInfo are the same, create and store only one of them


Code for timing:
```python
timer = Timer(
    stmt='a.view(-1)',
    setup='''\
import torch
a = torch.rand(4)''')

res = timer.collect_callgrind(repeats=2, number=10)[1]
```

Difference between master and this PR:
```
# Benchmark at master
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33be83690>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69286                      68442
    Baseline:             1332                       1188
10 runs per measurement, 1 thread

# Benchmark at this branch
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33bd7ec30>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69437                      68562
    Baseline:             1363                       1188
10 runs per measurement, 1 thread

# Difference between the two
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fe1216e9a00>
    160  ???:0x000000000a11c8d0
     60  torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta
     60  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
     40  ???:0x0000000008e14f50
     40  ???:0x0000000008e05bd0
     40  ???:0x0000000008e05480
     40  ???:0x0000000008e036d0
     40  ???:0x0000000008e02720
     30  make_variable_differentiable_view
    ...
    -20  ???:0x0000000008e02060
    -20  ???:0x0000000008e01fd0
    -30  ???:torch::autograd::isForwardADEnabled()
    -40  ???:0x0000000008e14f90
    -40  ???:0x0000000008e05c00
    -40  ???:0x0000000008e054a0
    -40  ???:0x0000000008e036f0
    -40  ???:0x0000000008e02740
   -160  ???:0x000000000a11d8d0

Total: 120

```


[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -531,6 +531,11 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
+  // Optimization to reduce the number of ViewInfo we create.
+  // In the (very common) case where backward_info_ == forward_info_
+  // we only populate backward_info_ and set shared_view_info_ = true
+  bool shared_view_info_;
```
What I wouldn't give for an algebraic data type here...

```
data ViewMeta
  = NoViewInfo
  | SharedViewInfo ViewInfo
  | DistinctViewInfo ViewInfo ViewInfo
```

ezyang replied:
The comment here would be clearer if you explicitly spelled out the invariants expected to be respected at all times on ViewInfo.

ezyang reviewed (approved):
okey dokey

albanD commented:
I am fixing the CI right now, but it will be only PRs below this one in the stack.

albanD committed:
Update on "Improve perf for forward AD view handling"


This PR performs optimization on the ViewInfo handling to remove the need for the "no forward AD mode".
- When the forward and backward ViewInfo are the same, create and store only one of them


Code for timing:
```python
timer = Timer(
    stmt='a.view(-1)',
    setup='''\
import torch
a = torch.rand(4)''')

res = timer.collect_callgrind(repeats=2, number=10)[1]
```

Difference between master and this PR:
```
# Benchmark at master
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33be83690>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69286                      68442
    Baseline:             1332                       1188
10 runs per measurement, 1 thread

# Benchmark at this branch
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33bd7ec30>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69437                      68562
    Baseline:             1363                       1188
10 runs per measurement, 1 thread

# Difference between the two
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fe1216e9a00>
    160  ???:0x000000000a11c8d0
     60  torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta
     60  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
     40  ???:0x0000000008e14f50
     40  ???:0x0000000008e05bd0
     40  ???:0x0000000008e05480
     40  ???:0x0000000008e036d0
     40  ???:0x0000000008e02720
     30  make_variable_differentiable_view
    ...
    -20  ???:0x0000000008e02060
    -20  ???:0x0000000008e01fd0
    -30  ???:torch::autograd::isForwardADEnabled()
    -40  ???:0x0000000008e14f90
    -40  ???:0x0000000008e05c00
    -40  ???:0x0000000008e054a0
    -40  ???:0x0000000008e036f0
    -40  ???:0x0000000008e02740
   -160  ???:0x000000000a11d8d0

Total: 120

```


[ghstack-poisoned]

albanD committed:
Update on "Improve perf for forward AD view handling"


This PR performs optimization on the ViewInfo handling to remove the need for the "no forward AD mode".
- When the forward and backward ViewInfo are the same, create and store only one of them


Code for timing:
```python
timer = Timer(
    stmt='a.view(-1)',
    setup='''\
import torch
a = torch.rand(4)''')

res = timer.collect_callgrind(repeats=2, number=10)[1]
```

Difference between master and this PR:
```
# Benchmark at master
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33be83690>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69286                      68442
    Baseline:             1332                       1188
10 runs per measurement, 1 thread

# Benchmark at this branch
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33bd7ec30>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69437                      68562
    Baseline:             1363                       1188
10 runs per measurement, 1 thread

# Difference between the two
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fe1216e9a00>
    160  ???:0x000000000a11c8d0
     60  torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta
     60  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
     40  ???:0x0000000008e14f50
     40  ???:0x0000000008e05bd0
     40  ???:0x0000000008e05480
     40  ???:0x0000000008e036d0
     40  ???:0x0000000008e02720
     30  make_variable_differentiable_view
    ...
    -20  ???:0x0000000008e02060
    -20  ???:0x0000000008e01fd0
    -30  ???:torch::autograd::isForwardADEnabled()
    -40  ???:0x0000000008e14f90
    -40  ???:0x0000000008e05c00
    -40  ???:0x0000000008e054a0
    -40  ???:0x0000000008e036f0
    -40  ???:0x0000000008e02740
   -160  ???:0x000000000a11d8d0

Total: 120

```


[ghstack-poisoned]

albanD committed:
Update on "Improve perf for forward AD view handling"


This PR performs optimization on the ViewInfo handling to remove the need for the "no forward AD mode".
- When the forward and backward ViewInfo are the same, create and store only one of them


Code for timing:
```python
timer = Timer(
    stmt='a.view(-1)',
    setup='''\
import torch
a = torch.rand(4)''')

res = timer.collect_callgrind(repeats=2, number=10)[1]
```

Difference between master and this PR:
```
# Benchmark at master
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33be83690>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69286                      68442
    Baseline:             1332                       1188
10 runs per measurement, 1 thread

# Benchmark at this branch
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fe33bd7ec30>
a.view(-1)
setup:
  import torch
  a = torch.rand(4)

                           All          Noisy symbols removed
    Instructions:        69437                      68562
    Baseline:             1363                       1188
10 runs per measurement, 1 thread

# Difference between the two
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fe1216e9a00>
    160  ???:0x000000000a11c8d0
     60  torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta
     60  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
     40  ???:0x0000000008e14f50
     40  ???:0x0000000008e05bd0
     40  ???:0x0000000008e05480
     40  ???:0x0000000008e036d0
     40  ???:0x0000000008e02720
     30  make_variable_differentiable_view
    ...
    -20  ???:0x0000000008e02060
    -20  ???:0x0000000008e01fd0
    -30  ???:torch::autograd::isForwardADEnabled()
    -40  ???:0x0000000008e14f90
    -40  ???:0x0000000008e05c00
    -40  ???:0x0000000008e054a0
    -40  ???:0x0000000008e036f0
    -40  ???:0x0000000008e02740
   -160  ???:0x000000000a11d8d0

Total: 120

```

Differential Revision: [D28071505](https://our.internmc.facebook.com/intern/diff/D28071505)

[ghstack-poisoned]
